File: lintcode/cpp/Best_Time_to_Buy_and_Sell_Stock.cpp
class Solution {
public:
    /**
     * @param prices: Given an integer array
     * @return: Maximum profit
     */
    int maxProfit(vector<int> &prices) {
        // write your code here
        if(prices.size() <= 1) return 0;
        
        int n = prices.size();
        vector<int> diff(n-1, 0);
        for(int i = 1; i < n; i++){
            diff[i-1] = prices[i] - prices[i-1];
        }
        
        return maxSubSum(diff);
    }
    
private:
    int maxSubSum(vector<int> &A){
        
        int n = A.size();
        vector<int> s(n, 0);
        s[0] = max(0, A[0]);
        int maxSum = s[0];
        
        for(int i = 1; i < n; i++){
            s[i] = max(0, max(s[i-1]+A[i], A[i]));
            if(s[i] > maxSum) maxSum = s[i];
        }
        return maxSum;
    }
};

File: lintcode/java/BestTimetoBuyandSellStock.java
public class Solution {
    /**
     * @param prices: Given an integer array
     * @return: Maximum profit
     */
    public int maxProfit(int[] prices) {
        if(prices.length <= 1) return 0;
        
        int n = prices.length;
        int[] diff = new int[n-1];
        for(int i = 1; i < n; i++){
            diff[i-1] = prices[i] - prices[i-1];
        }
        
        return maxSum(diff);
    }
    
    private int maxSum(int[] A){
        int n = A.length;
        int[] t = new int[n+1];
        t[0] = 0;
        
        int maxSum = 0;
        for(int i = 1; i <= n; i++){
            t[i] = Math.max(t[i-1]+A[i-1], Math.max(A[i-1], 0));
            if(t[i] > maxSum){
                maxSum = t[i];
            }
        }
        
        return maxSum;
    }
}
